### ISSUE TITLE
GO_SRCROOT and S

### ISSUE THREAD

**Issue by roman-khimov** [madisongh/meta-golang#6]
At the moment golang.bbclass sets S for its recipes this way:

```
S_GOROOT = "${WORKDIR}/go"
S = "${S_GOROOT}/src"
```

But the actual sources are unpacked into `${S}/${GO_SRCROOT}`, so when you need to do something in the sources you have to use `${S}/${GO_SRCROOT}`, that includes `LIC_FILES_CHKSUM` file references, any appends/prepends to `do_compile()` or `do_configure()` and especially annoying is the need to add `patchdir=${GO_SRCROOT}` for patches in the `SRC_URI`.

I think it goes against traditional OE `${S}` definition. To build the package we only need `GOPATH` to be set up properly, if I understand it correctly, so we should change `S` to point into `${S_GOROOT}/src/${GO_SRCROOT}` and fix current `${GO_SRCROOT}` (mis)uses.

**Comment by madisongh** [comment]
There were some good reasons for this when I originally set this up, mainly to make it easier to handle having multiple packages' sources available when there were circular dependencies between them.  Now that the `go` tool has a well-defined mechanism for "vendoring" in needed packages, it's probably worth revisiting this to see if the variable setup could be simplified.

**Comment by madisongh** [comment]
I have looked at this again, and unfortunately I think we're stuck with this un-OE-like arrangement.  Other approaches that I've tried to put `${S}` at the top of the package's source, as requested, run into other issues.

The problems stem from trying support vendoring (including applying patches to vendored packages, if needed) and use of the `go` tool's `...` wildcard to build all of a given go package's contents (without creating binaries for the vendored packages).  If someone can find a workable solution, I'd be willing to consider those changes.
